Updated parser code

NikhilPappu opened the pull request:
The major changes in this commit include the code I have changed
in _listener_autolev_antlr.py. The changes to other files are
minor. I have also made the changes requested in PR #14758
after it had been merged.

Some specific changes are:
1. Changed the input rule in the grammar and parser code to fix errors.
2. Added a flag include_pydy in parse_autolev.
3. Changed the doctest in `__init__.py` to make it look better.
4. Removed the print option. stdout is now the default.
5. Made various changes to _listener_autolev_antlr to parse
   more files. Revamped the processVariables function quite a bit.
   Changed the mass function and the pydy output code a bit followed
   by some minor changes.
6. I have also added a .subs(kindiffdict()) in the forcing full method
   of kane.py. This is required for the pydy numerical code to work in
   some cases. This doesn't break any tests.

<!-- BEGIN RELEASE NOTES -->
- parsing
   - Made changes to _listener_autolev_antlr in parsing.autolev. Revamped the processVariables function 
     quite a bit. Changed the mass function and the pydy output code a bit followed
     by some minor changes. Made some minor changes to other files in parsing.autolev.
   - Also changed some code in tests/test_autolev.py (Added commented test code for the GitLab repo 
     tests and changed zip to zip_longest which is more correct.
 
- physics.mechanics
   - Added a .subs(kindiffdict()) in the forcing full method of kane.py. This is required for the pydy 
     numerical code to work in some cases. This doesn't break any tests.
<!-- END RELEASE NOTES -->

NikhilPappu committed:
Updated the parser code and made changes requested in #14758.

The major changes in this commit include the code I have changed
in _listener_autolev_antlr.py. The changes to other files are
minor. I have also made the changes requested in PR #14758
after it had been merged.

Some specific changes are:
1. Changed the input rule in the grammar and parser code to fix errors.
2. Added a flag include_pydy in parse_autolev.
3. Changed the doctest in __init__.py to make it look better.
4. Removed the print option. stdout is now the default.
5. Made various changes to _listener_autolev_antlr to parse
   more files. Revamped the processVariables function quite a bit.
   Changed the mass function and the pydy output code a bit followed
   by some minor changes.
6. I have also added a .subs(kindiffdict()) in the forcing full method
   of kane.py. This is required for the pydy numerical code to work in
   some cases. This doesn't break any of the test cases.
7. Changed zip to zip_longest in test_autolev.py. Also added commented
   code for the tests in the GitLab repo.

moorepants commented:
@NikhilPappu Please fill out the pull request template so we know what the PR is about. This PR also looks very large, like your last one. It is much easier for us to review if you break it into atomic changes spread across multiple PRs. It will speed up review if you do that.

asmeurer commented:
Looks like you found another bug in the bot. Thanks @NikhilPappu!

asmeurer commented:
Also please write real commit messages. "Updated code" could literally apply to any commit. The commit message should describe what the commit changes in a way that someone who comes across it in a year from now could understand what was changed and, importantly, why.

moorepants commented:
@NikhilPappu Here is a nice guide to writing good commit messages: https://chris.beams.io/posts/git-commit/

asmeurer commented:
We also have a guide in our own development workflow https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages. I might take a look at the guide Jason posted and see if our guide can be improved.

asmeurer commented:
Also one thing that I would add that doesn't seem to be in either guide is to never use the `-m` flag to `git commit`. Let it open your editor so you can write a real message. Using `-m` nudges you to writing messages that are too short. If you don't like how `git commit` opens vim when using `-m` you can configure your `EDITOR` environment variable to open something better (`nano` is the simplest option, and most GUI editors have a command line tool that you can set `EDITOR` to that will open in them).

NikhilPappu commented:
@asmeurer @moorepants 
Sorry about the not providing any descriptions. I made the PR in a hurry and just wanted to check the Travis build for once. I will write the PR message and the release notes.
As for multiple PRs, I think the changes made here are minor for the most part. I think I could open new PRs for the tests added to test_functions.py and test_frame.py in physics/mechanics and vector.
As for the commit message I will change it to a more descriptive one. It's hard to say much in just the title so I shall give a proper description in the body of the commit message.

NikhilPappu committed:
Changed parse_autolev api

NikhilPappu committed:
Added a warning for Python keywords conflict

NikhilPappu committed:
Merge branch 'master' into autolev_parser

NikhilPappu commented:
@moorepants I have changed the `parse_autolev` api and added a warning for the keywords as per your suggestion in the documentation PR.

moorepants commented:
@NikhilPappu resolve the merge conflicts.

NikhilPappu committed:
Merge branch 'master' into autolev_parser

NikhilPappu committed:
Fixed an error in setup.py

NikhilPappu commented:
@moorepants I resolved them.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -1,56 +1,60 @@
-         2. Can be a string containing Autolev code.
+    =========
+    autolev_code : str, any object with a readlines() method (such as a file handle or StringIO)
+        Autolev code...
```
This is a dangling sentence, make complete.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -1,56 +1,60 @@
+    INPUT Q1=.1,Q2=.2,U1=0,U2=0
+    INPUT TFINAL=10, INTEGSTP=.01
+    CODE DYNAMICS() some_filename.c
+    ------------FILE--------------
```
Does this render correctly when parsed? Don't you need a `::` or `.. code: XXX`?

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -1,56 +1,60 @@
-    >>> l.append("INPUT TFINAL=10, INTEGSTP=.01)"  # doctest: +SKIP
-    >>> l.append("CODE DYNAMICS() double_pendulum.c)"  # doctest: +SKIP
-    >>> parse_autolev("\\n".join(l), "print")  # doctest: +SKIP
+    >>> print(parse_autolev(open("double_pendulum.al"), include_numeric=True))  # doctest: +SKIP
```
Should use:
```python
with open("double_pendulum.al", 'r') as f:
    parse_autolev(f, include_numeric=True)
```

moorepants replied:
Don't skip the doctest if possible.

Load the text like:

```pycon
>>> my_al_text = """\
stuff
more stuff
another thing"""
```

NikhilPappu replied:
I have changed the doctest and the test passes locally but Travis won't pass it because of some antlr import stuff.

moorepants replied:
Does antlr need to be added to the `.travis.yml` file?

moorepants replied:
Is it because antlr is not installed in the doc tests section on travis?

moorepants replied:
Open an issue for this "antlr not available for doctests".

asmeurer replied:
Use the `doctests_depends_on` decorator to skip the doctest if a module isn't installed.

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -1,56 +1,60 @@
+
     q1, q2, u1, u2 = me.dynamicsymbols('q1 q2 u1 u2')
     q1d, q2d, u1d, u2d = me.dynamicsymbols('q1 q2 u1 u2', 1)
     l, m, g=sm.symbols('l m g', real=True)
```
space around equals, please run PEP8 linter on all code

moorepants commented on sympy/parsing/autolev/test-examples/pydy-example-repo/non_min_pendulum.py:
```diff
@@ -33,3 +34,7 @@
+# Works
+# Numerical parts for such kind of things won't work.
+# The current numerical parts only work with the simple kane's method use (no_args),
+# no config constraints etc. Have to add auxiliary stuff later.
```
An issue should be opened about supporting config constraints, auxiliary speeds, etc.

NikhilPappu replied:
Auxiliary speeds have to be added and I will open an issue for that.
The problem with config constraints is that they aren't formally specified in Autolev unlike motion constraints. So, they cannot be determined for the Kane's method object.
Can we ask the users to use a made up command like ConfigConstraint() or something?

NikhilPappu replied:
Configuration constraint issue: #15172.

moorepants reviewed (changes requested):
Some minor comments added. I'm going to also test this locally and will give any feedback if needed.

moorepants commented:
First error I got when trying locally is:

```
/home/moorepants/src/sympy/sympy/external/importtools.py:145: UserWarning: antlr4 module is not installed
  warnings.warn("%s module is not installed" % module, UserWarning)
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-3-7dc7f7398af9> in <module>()
      1 with open('whipple.al', 'r') as f:
----> 2     res = parse_autolev(f)
      3 

~/src/sympy/sympy/parsing/autolev/__init__.py in parse_autolev(autolev_code, include_numeric)
     97 
     98     if _autolev is not None:
---> 99         return _autolev.parse_autolev(autolev_code, include_numeric)

~/src/sympy/sympy/parsing/autolev/_parse_autolev_antlr.py in parse_autolev(autolev_code, include_numeric)
     17     antlr4 = import_module('antlr4', warn_not_installed=True)
     18     if not antlr4:
---> 19         raise ImportError("Autolev parsing requires the antlr4 python package,"
     20                           " provided by pip (antlr4-python2-runtime or"
     21                           " antlr4-python3-runtime) or"

ImportError: Autolev parsing requires the antlr4 python package, provided by pip (antlr4-python2-runtime or antlr4-python3-runtime) or conda (antlr-python-runtime)
```

I then tried:

```
$ conda search antlr
moorepants@garuda:~$ conda search antlr
Loading channels: done
# Name                  Version           Build  Channel             
antlr                     2.7.7               2  conda-forge         
antlr                     2.7.7               3  conda-forge         
antlr                     2.7.7               4  conda-forge         
antlr                     2.7.7               5  conda-forge         
antlr                     2.7.7               6  conda-forge         
antlr                     2.7.7               7  conda-forge         
antlr                     2.7.7      hfc679d8_7  conda-forge         
antlr                     2.7.7          py27_0  conda-forge         
antlr                     2.7.7          py27_1  conda-forge         
antlr                     4.7.1               0  conda-forge
```

There does not seem to be any package called `antlr-python-runtime` in the conda channels defaults of conda-forge.

moorepants commented:
I take that back. It is just conda's poor search command:

```
moorepants@garuda:~$ conda search antlr*
Loading channels: done
# Name                  Version           Build  Channel             
antlr                     2.7.7               2  conda-forge         
antlr                     2.7.7               3  conda-forge         
antlr                     2.7.7               4  conda-forge         
antlr                     2.7.7               5  conda-forge         
antlr                     2.7.7               6  conda-forge         
antlr                     2.7.7               7  conda-forge         
antlr                     2.7.7      hfc679d8_7  conda-forge         
antlr                     2.7.7          py27_0  conda-forge         
antlr                     2.7.7          py27_1  conda-forge         
antlr                     4.7.1               0  conda-forge         
antlr-python-runtime           4.7.1          py27_0  conda-forge         
antlr-python-runtime           4.7.1          py35_0  conda-forge         
antlr-python-runtime           4.7.1          py36_0  conda-forge         
```

NikhilPappu commented:
@moorepants I'll make the changes shortly. The whipple.al file should work (other than the linearization part) if the errors are fixed. I'll look into this when I find time.

moorepants commented on sympy/physics/mechanics/kane.py:
```diff
@@ -638,7 +638,7 @@ def forcing_full(self):
             raise ValueError('Need to compute Fr, Fr* first.')
         f1 = self._k_ku * Matrix(self.u) + self._f_k
-        return -Matrix([f1, self._f_d, self._f_dnh])
+        return -Matrix([f1.subs(self.kindiffdict()), self._f_d.subs(self.kindiffdict()), self._f_dnh.subs(self.kindiffdict())])
```
I think `xreplace()` should, in general, be used instead of subs.

moorepants replied:
This also needs a unit test.

moorepants replied:
I'm also not sure that this is right place to do this. It could have quite the affect on speed of computation.

This is the kind of thing that needs to be in a separate PR.

NikhilPappu replied:
I shall use xreplace and add the test. There doesn't seem to be any difference in the runtime of the current mechanics tests though.

moorepants replied:
This gets exacerbated on larger problems.

NikhilPappu replied:
I am not sure about a workaround. I shall move this to another PR then.

moorepants commented on sympy/parsing/tests/test_autolev.py:
```diff
@@ -11,36 +13,55 @@ def _test_examples(in_filename, out_filename, test_name=""):
+#     for i in l:
+#         in_filepath = os.path.join("autolev-tutorial", i + ".al")
+#         out_filepath = os.path.join("autolev-tutorial", i + ".py")
+#         _test_examples(in_filepath, out_filepath, i)
```
Don't leave commented code in like this.

NikhilPappu replied:
I left these like this because the autolev-tutorial and dynamics-online directories can be copied over from the GitLab repo and tested by uncommenting this code.

moorepants replied:
I see. It would be better to have them uncommented and use an if statement that runs them if the files are there, but doesn't if the files aren't.

NikhilPappu replied:
That sounds better. I shall change it.

NikhilPappu committed:
Made the requested changes

Also fixed the position vector error observed in #15165.

NikhilPappu commented:
@moorepants I have addressed the comments and made some changes. I have also fixed the position vector error observed in #15165. I think this PR should be almost ready for merging now. As for whipple.al, I will have to run it line by line and check the reasons for the errors. I shall do this later.

NikhilPappu committed:
Removed kane.py changes and replaced the commented code

moorepants commented on sympy/parsing/autolev/__init__.py:
```diff
@@ -19,55 +18,51 @@ def parse_autolev(autolev_code, include_numeric=False):
+    ... INPUT M=1,G=9.81,L=1\\n\\
+    ... INPUT Q1=.1,Q2=.2,U1=0,U2=0\\n\\
+    ... INPUT TFINAL=10, INTEGSTP=.01\\n\\
+    ... CODE DYNAMICS() some_filename.c"
```
Is the `\\n\\` the only way to get this work? Makes ti quite ugly...

You can do this too:

```pycon
>>> my_text = ('first line',
... 'second line',
... 'third line')
>>> my_text = '\n'.join(my_text)
```

Maybe that is a little cleaner??

moorepants commented:
Looks like you need something that says "if using symengine" use subs, otherwise use xreplace.:

```
___________ sympy/physics/mechanics/tests/test_kane.py:test_one_dof ____________
Traceback (most recent call last):
  File "sympy/utilities/runtests.py", line 1277, in _timeout
    function()
  File "/home/travis/build/sympy/sympy/sympy/physics/mechanics/tests/test_kane.py", line 40, in test_one_dof
    KM.mass_matrix_full.LUsolve(KM.forcing_full)) == zeros(2, 1)
  File "sympy/physics/mechanics/kane.py", line 641, in forcing_full
    return -Matrix([f1.xreplace(self.kindiffdict()), self._f_d.xreplace(self.kindiffdict()), self._f_dnh.xreplace(self.kindiffdict())])
AttributeError: 'symengine.lib.symengine_wrapper.MutableDenseMatrix' object has no attribute 'xreplace'
```

moorepants commented:
Opened issue here: https://github.com/symengine/symengine/issues/1456

NikhilPappu committed:
Cleaned up the doctest

NikhilPappu committed:
Added doctest_depends_on decorator

NikhilPappu commented:
@moorepants I think this can be merged now.

moorepants commented:
Great!
